File: bicycle/bicycle.py
from transportation import Transportation
from typing import Sequence,Set,List,Tuple,Dict
from wheel import Wheel
from transportation import Transportation

class Bicycle(Transportation):
    def __init__(self, color : str, diameter : float = Wheel.DEFAULT_DIAMETER):
        super(Bicycle,self).__init__(started = False)
        self._wheels : List[Wheel] = [Wheel(diameter),Wheel(diameter)]
        self._color : str = color

    @property
    def front(self) -> Wheel:
        return self._wheels[0]

    @property
    def rear(self) -> Wheel:
        return self._wheels[1]

    @property
    def flat(self) -> bool:
        return self.front.flat or self.rear.flat

# override
    @property
    def going(self):
        return  self.started and not self.flat

File: cpp-bicycle/src/Bicycle.cpp
#include <stdexcept>
#include "Bicycle.h"
namespace go
{

Bicycle::Bicycle(const std::string &_color, double _diameter)
    : Transportation(false), color(_color), wheels(2)
{
  wheels[0] = std::shared_ptr<Wheel>(new Wheel(_diameter));
  wheels[1] = std::shared_ptr<Wheel>(new Wheel(_diameter));
}

const std::string &Bicycle::getColor() const { return color; }
void Bicycle::setColor(const std::string &value) { color = value; }

Wheel &Bicycle::getFront() { return *wheels[0]; }
const Wheel &Bicycle::getFront() const { return *wheels[0]; }

Wheel &Bicycle::getRear() { return *wheels[1]; }
const Wheel &Bicycle::getRear() const { return *wheels[1]; }

bool Bicycle::isFlat() const { return getFront().isFlat() || getRear().isFlat(); }

bool Bicycle::isGoing() const { return isStarted() && !isFlat(); }

} // namespace go
